src/metadata/firmware_download.rs

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+// SPDX-FileCopyrightText: 2025 1BitSquared <info@1bitsquared.com>
+// SPDX-FileContributor: Written by Rachel Mant <git@dragonmux.network>
+// SPDX-FileContributor: Modified by P-Storm <pauldeman@gmail.com>
+
+use std::ffi::OsStr;
+use std::path::PathBuf;
+
+use color_eyre::eyre::eyre;
+use color_eyre::Result;
+use serde::Deserialize;
+use url::Url;
+
+#[derive(Deserialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct FirmwareDownload
+{
+	#[serde(rename = "friendlyName")]
+	pub friendly_name: String,
+	#[serde(rename = "fileName")]
+	pub file_name: PathBuf,
+	pub uri: Url,
+}
+
+impl FirmwareDownload
+{
+	pub fn build_documentation_url(&self) -> Result<Url>
+	{
+		// Convert the path compoment of the download URI to a Path
+		let mut docs_path = PathBuf::from(&self.uri.path());
+		if docs_path.extension() == Some(OsStr::new("elf")) {
+			docs_path.set_extension("md");	
+		} else {
+			return Err(eyre!("Suspected"));
+		}
+
+		// Copy only the origin
+		let mut docs_uri = self.uri.clone();
+		docs_uri.set_path(docs_path.to_str().expect("Can't set a path from a doc path"));
+		docs_uri.set_fragment(None);
+		docs_uri.set_query(None);
+
+		Ok(docs_uri)
+	}
+
+	pub fn build_release_uri(&self, release: &str) -> Result<Url>
+	{
+		// Expected uri input: https://github.com/blackmagic-debug/blackmagic/releases/download/<release>/blackmagic-native-v2_0_0-rc1.elf
+		let release_position_option = self.uri.path_segments()
+			.expect("Can't be base")
+			.position(|r| r == release);
+		
+		// The download position should be before the release position
+		let download_position = match release_position_option {
+			Some(position) => position.checked_sub(1).ok_or_else(|| eyre!("The release segment '{}' can't be the first one", release))?,
+			None => return Err(eyre!("The provided uri doesn't contain the release segment '{}'", release)) 
+		};
+		
+		// Take the segments up to the /download/ position
+		// From: /blackmagic-debug/blackmagic/releases/download/<release>/blackmagic-native-v2_0_0-rc1.elf
+		// To  : /blackmagic-debug/blackmagic/releases/tag/<release>
+		let release_path: PathBuf = self.uri.path_segments()
+			.expect("Can't be base")
+			.take(download_position)
+			// Now add on /tag/ and the release number
+			.chain(["tag", release])
+			.collect();
+		
+		let new_path = release_path.to_str()
+			.expect("cannot be base");
+		
+		// build on the premise that the shape is smaller, or at least the segment "tag" is smaller then "download"
+		if self.uri.path().len() < new_path.len() {
```

Comment:
We shouldn't need this check if we keep the top ones, so shall we keep this check in place or remove this @dragonmux

Reply:
Given the other checks, you can remove this. Alternatively you can keep this and remove the other checks - permission vs forgiveness. Both result in the same basic effect here.. though arguably the permission checks are more thorough about validating assumptions so. Let's keep those and ditch this one.